### PR TITLE
refactor(attention): use fixed page tables for paged block-sparse attention

### PR DESCRIPTION
**Pull request by heyuhhh**
## 📌 Description

Make both paged PrimTS block-sparse entry points, `BlockSparsePagedTSWrapper.run()` and the one-shot
`block_sparse_attention_with_paged_kv_cache()`, consume a fixed 2D `block_tables` tensor with an explicit row
stride instead of CSR page indices (`paged_kv_indptr` / `paged_kv_indices`). This matches the PrimTS decode
contract from #4829, so callers that already hold page tables (TensorRT-LLM's paged KV cache manager, for example)
can pass a zero-copy view instead of converting on the host. The reusable wrapper reads the table directly during
`run()`; the one-shot API validates the live prefix of every page-table row on the device before planning. Like the
dense PrimTS wrappers, `run()` accepts `validate=False` to treat every argument as a trusted binding and skip the
explicit structural, plan-geometry, and alias checks.

The PrimTS block-sparse trace templates are rewritten as literal schemas in two factories, one per K/V storage
form, instead of deriving the paged and wrapper templates from another template by deleting or rewriting entries.
Template names, dispatchers, and the axis, input, constraint, and tag sets are unchanged apart from the page-table
and `validate` inputs; the six block-sparse goldens are regenerated and a test now pins every golden to its
template.

Follow-up to #4872 and #5002; the contiguous entry points are untouched.

## 🔍 Related Issues

Follow-up to #4829 (PrimTS decode page-table contract), #4872 and #5002 (PrimTS block-sparse attention).

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

`tests/attention/test_attention_ts_block_sparse.py` (paged one-shot and wrapper cases now build page tables,
including padded row strides and live-prefix validation) and `tests/trace/test_fi_trace_template_consistency.py`
(golden-to-template pinning): 1056 passed, 1 skipped on B200 (SM100) with CUTLASS DSL 4.7.0; pristine `main` gives 1058 passed, 1 skipped in the same environment (the difference is the reshaped paged test cases).

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

The paged API signature changes: `paged_kv_indptr` / `paged_kv_indices` are replaced by `block_tables` (Int32
`[B, C]`, padded outer stride allowed) plus `seq_lens_kv`, and `run()` gains `validate=True`. The README documents
the new value contract. TensorRT-LLM already consumes this form through its vendored PrimTS copy.

## Summary by CodeRabbit

- **Changed**
  - Paged block-sparse attention now uses fixed 2D `block_tables` with optional row padding for KV-cache page addressing, replacing the previous indirection arrays.
  - Run methods support optional structural and aliasing validation, including a faster trusted-input path when disabled.
  - One-shot APIs now accept sequence lengths positionally.

- **Bug Fixes**
  - Validation now checks page capacity, tensor structure, device compatibility, aliasing, and physical page IDs with clearer errors.

- **Tests**
  - Expanded coverage for block-table validation, tracing schemas, and paged block-sparse correctness.

### CODE REVIEW

**Comment by qsang-nv** [comment]
**Two loose ends after the CSR → `block_tables` migration.**

`BlockSparsePagedTSWrapper.plan()`'s docstring (`flashinfer/attention/prims_ts/block_sparse.py:564-565`) still says "Every run supplies live page-table offsets, page IDs, sequence lengths, ...", describing the removed `paged_kv_indptr`/`paged_kv_indices` CSR contract. `run()`'s own docstring and the README were updated to describe `block_tables`, but `plan()`'s was not.

`test_prims_ts_block_sparse_goldens_match_templates` (`tests/trace/test_fi_trace_template_consistency.py:803-824`) only compares each golden's input `name`, `optional`, and `description` against the template. It never compares input `shape`, and never touches `outputs` (name/shape/dtype/param) at all. A golden with a mis-bound tensor shape or wrong output ABI would still pass this test.

File: tests/attention/test_attention_ts_block_sparse.py [hunk 1]
```diff
@@ -5517,16 +5626,18 @@ def varlen_reference(
 
     block_indptr = indptr_a.clone()
     block_indices = indices_a.clone()
-    paged_kv_indices = page_ids_a.clone()
-    paged_kv_indptr = paged_kv_indptr_a.clone()
+    block_table_storage = torch.full(
+        (case.batch_size, 3), -1, device="cuda", dtype=torch.int32
+    )
+    block_tables = block_table_storage[:, :2]
+    block_tables.copy_(torch.stack((page_ids_a[:2], page_ids_a[1:3])))
```
[the remark this answers is not in the document]
Comment (heyuhhh) [reply tests/attention/test_attention_ts_block_sparse.py]: The rows are intentional. This test uses page_size=64; in round A request 0 holds 48 live tokens and in round B request 1 holds 64, so each owns a single live page. block_tables documents that entries beyond the live length are ignored, so the trailing entries [0, 2] and [2, 2] exercise exactly that rule while the live prefixes [0], [2, 3], [1, 4], [2] match the references. Replacing them with the reference page IDs would drop that coverage. The rows are now spelled out as literals with a comment explaining this.

**Comment by heyuhhh** [comment]
@qsang-nv thanks for the review! The docs and validations have made up in commit 61f8848864434c175ccbed63fb997bcc21e2d2d0

File: flashinfer/attention/prims_ts/block_sparse.py [hunk 2]
```diff
@@ -698,28 +731,26 @@ def run(
 def block_sparse_attention_with_paged_kv_cache(
```
[the remark this answers is not in the document]
Comment (heyuhhh) [reply flashinfer/attention/prims_ts/block_sparse.py:730]: It's an intentional change

File: flashinfer/attention/prims_ts/block_sparse.py [hunk 3]
```diff
@@ -815,6 +849,12 @@ def block_sparse_attention_with_paged_kv_cache(
         kv_dtype=k_cache.dtype,
         output_dtype=q.dtype if out is None else out.dtype,
     )
+    if table_capacity * page_size < static.seq_len_kv:
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/block_sparse.py]: those checks seem duplicated as shown in flashinfer/attention/prims_ts/_block_sparse/runtime.py

**Review by PerkzZheng — APPROVED** [review]

**Review by qsang-nv — APPROVED** [review]

File: flashinfer/attention/prims_ts/block_sparse.py (hunk 3 above)
Comment (heyuhhh) [reply flashinfer/attention/prims_ts/block_sparse.py]: Thanks for the findings! We can not directly remove these checks but can hoist it as a common helper. if `wrapper.run()` with `validate=True` it actually happens twice. However, they are in two different scopes: one is in one-shot API for the following inspect kernel and another is in `wrapper.run(..., validate=True)`

Comment (heyuhhh) [reply flashinfer/attention/prims_ts/block_sparse.py]: Updated in bded04a57885005e6c5a4b37dea7e01cfa3310dc

**Comment by bkryu** [comment]
@flashinfer-bot run

**Comment by heyuhhh** [comment]
@qsang-nv @bkryu There are some conflicts and i have resolved. Could you please help trigger the CI again? Thanks!

**Review by yuxianq — APPROVED** [review]
